File: 算法竞赛进阶指南/0x10基本数据结构/AcWing41.cpp
class MinStack {
  stack<int> a, b; // a中存原始，b中存栈底开头的每段最小值
public:

  void push(int x) {
      a.push(x);
      b.empty() ? b.push(x) : b.push(min(b.top(), x));
  }
  
  void pop() {
      a.pop(), b.pop();
  }
  
  int top() {
      return a.top();
  }
  
  int getMin() {
      return b.top();
  }
};

File: 0x10 基本数据结构/0x11 栈/AcWing41.py
class MinStack(object):

    def __init__(self):
        self.a, self.b = [], []

    def push(self, x):
        self.a.append(x)
        self.b.append(x) if not self.b else self.b.append(min(self.b[-1], x))

    def pop(self):
        self.a.pop()
        self.b.pop()

    def top(self):
        return self.a[-1]

    def getMin(self):
        return self.b[-1]